io/zenoh-links/zenoh-link-shm/src/unicast.rs

```diff
@@ -0,0 +1,439 @@
+            let mut invitation = [0; 4];
+            uplink.read_exact(&mut invitation).await?;
+            if invitation != PIPE_INVITATION {
+                bail!("Unexpected invitation received during pipe handshake!")
+            }
+
+            // create downlink
+            let mut downlink = PipeW::new(&path_downlink).await?;
+            // echo invitation to other party
+            downlink.write_all(&invitation).await?;
+
+            // send newly established link to manager
+            manager
+                .send_async(LinkUnicast(Arc::new(UnicastPipe {
+                    r: async_std::sync::RwLock::new(uplink),
+                    w: async_std::sync::RwLock::new(downlink),
+                    local,
+                    remote,
+                })))
+                .await?;
+
+            ZResult::Ok(())
+        });
+
+        Ok(Self {
+            listening_task_handle,
+            uplink_locator: c_local,
+        })
+    }
+
+    async fn stop_listening(self) {
+        self.listening_task_handle.cancel().await;
+    }
+}
+
+struct UnicastPipeClient;
+impl UnicastPipeClient {
+    async fn connect_to(endpoint: EndPoint) -> ZResult<UnicastPipe> {
+        let (path_uplink, path_downlink, access_mode) = parse_pipe_endpoint(&endpoint);
+
+        // create Locators
+        let local = Locator::new(
+            endpoint.protocol(),
+            path_downlink.as_str(),
+            endpoint.metadata(),
+        )?;
+        let remote = Locator::new(
+            endpoint.protocol(),
+            path_uplink.as_str(),
+            endpoint.metadata(),
+        )?;
+
+        // create uplink
+        let mut uplink = PipeW::new(&path_uplink).await?;
+        // create downlink
+        let mut downlink = PipeR::new(&path_downlink, access_mode).await?;
+
+        // write invitation
+        uplink.write_all(PIPE_INVITATION).await?;
+
+        // read responce
+        let mut responce: [u8; 4] = [0, 0, 0, 0];
+        downlink.read_exact(&mut responce).await?;
+        // check responce
+        if responce != PIPE_INVITATION {
+            bail!("Unexpected responce received during pipe handshake!")
+        }
+
+        Ok(UnicastPipe {
+            r: async_std::sync::RwLock::new(downlink),
+            w: async_std::sync::RwLock::new(uplink),
+            local,
+            remote,
+        })
+    }
+}
+
+struct UnicastPipe {
+    r: async_std::sync::RwLock<PipeR>,
+    w: async_std::sync::RwLock<PipeW>,
+    local: Locator,
+    remote: Locator,
+}
+impl Drop for UnicastPipe {
+    fn drop(&mut self) {}
+}
+#[async_trait]
+impl LinkUnicastTrait for UnicastPipe {
+    async fn close(&self) -> ZResult<()> {
+        log::trace!("Closing SHM Pipe link: {}", self);
+        Ok(())
+    }
+
+    async fn write(&self, buffer: &[u8]) -> ZResult<usize> {
+        //todo: I completely don't like the idea of locking reader\writer each time, but LinkUnicastTrait requires self.r and self.w to be Sync
```

Comment:
By construction in the rest of zenoh logic there is only one task writing at the time. So, an `UnsafeCell` could also do the trick.

Reply:
In this case, why not relaxing Sync requirement in `LinkUnicastTrait` ?

Reply:
It can be investigated. This trait was defined at the very beginning (it was one of the first pieces of zenoh). So, the whole link trait could be revised. But that's I believe is out-of-scope of this PR.

Reply:
https://zettascale.atlassian.net/browse/ZEN-90